core/distributed/matrix.cpp

```diff
@@ -0,0 +1,440 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/matrix.hpp>
+
+
+#include <ginkgo/core/distributed/vector.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/distributed/matrix_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace matrix {
+namespace {
+
+
+GKO_REGISTER_OPERATION(build_diag_offdiag,
+                       distributed_matrix::build_diag_offdiag);
+
+
+}  // namespace
+}  // namespace matrix
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm)
+    : Matrix(exec, comm, with_matrix_type<gko::matrix::Csr>())
+{}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    const LinOp* inner_matrix_type)
+    : Matrix(exec, comm, inner_matrix_type, inner_matrix_type)
+{}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    const LinOp* local_matrix_template, const LinOp* non_local_matrix_template)
+    : EnableLinOp<
+          Matrix<value_type, local_index_type, global_index_type>>{exec},
+      DistributedBase{comm},
+      send_offsets_(comm.size() + 1),
+      send_sizes_(comm.size()),
+      recv_offsets_(comm.size() + 1),
+      recv_sizes_(comm.size()),
+      gather_idxs_{exec},
+      non_local_to_global_{exec},
+      one_scalar_{},
+      local_mtx_{local_matrix_template->clone(exec)},
+      non_local_mtx_{non_local_matrix_template->clone(exec)}
+{
+    GKO_ASSERT(
+        (dynamic_cast<ReadableFromMatrixData<ValueType, LocalIndexType>*>(
+            local_mtx_.get())));
+    GKO_ASSERT(
+        (dynamic_cast<ReadableFromMatrixData<ValueType, LocalIndexType>*>(
+            non_local_mtx_.get())));
+    one_scalar_.init(exec, dim<2>{1, 1});
+    one_scalar_->fill(one<value_type>());
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Matrix<ValueType, LocalIndexType, GlobalIndexType>::convert_to(
+    Matrix<next_precision<value_type>, local_index_type, global_index_type>*
+        result) const
+{
+    GKO_ASSERT(this->get_communicator().size() ==
+               result->get_communicator().size());
```

Comment:
I think we should maybe copy the communicator object as well ? Currently, if the matrix being moved into has not been created with a communicator, then these `copy_from` and `move_to` operations throw. The default constructor creates it with MPI_COMM_WORLD, so I am also not completely sure about that.

Reply:
The idea was to handle the communicator similar to how we handle the executor, i.e. it doesn't change during assignments. I don't think it's possible to create a matrix without a communicator (unless we talk about MPI_COMM_NULL), so that shouldn't cause issues. But if the sizes differ than I throw here, because the whole send_sizes_, etc are not compatible.